src/charm.py

```diff
@@ -74,25 +79,133 @@ def _charmed_mlflow_layer(self) -> Layer:
+                f"as expected. Caught exception: '{str(e)}'",
+                BlockedStatus,
+            )
+
+        return obj_storage
+
+    def _get_env_vars(self, relational_db_data, object_storage_data):
+        """Return environment variables based on model configuration."""
+
+        ret_env_vars = {
+            "AWS_ENDPOINT_URL": f"http://{object_storage_data['service']}.{object_storage_data['namespace']}:{object_storage_data['port']}",  # noqa: E501
+            "AWS_ACCESS_KEY_ID": object_storage_data["access-key"],
+            "AWS_SECRET_ACCESS_KEY": object_storage_data["secret-key"],
+            "USE_SSL": str(object_storage_data["secure"]).lower(),
+            "DB_ROOT_PASSWORD": relational_db_data["root_password"],
+            "MLFLOW_TRACKING_URI": f"mysql+pymysql://root:{relational_db_data['root_password']}@{relational_db_data['host']}:{relational_db_data['port']}/{relational_db_data['database']}",  # noqa: E501
+        }
+        return ret_env_vars
+
+    def _validate_default_s3_bucket(self, obj_storage_data):
+        """Validates the default S3 store, ensuring bucket is accessible and creating if needed."""
+        # Validate the bucket name
+        bucket_name = self.config["default_artifact_root"]
+        if not validate_s3_bucket_name(bucket_name):
+            msg = (
+                f"Invalid value for config default_artifact_root '{bucket_name}'"
+                f" - value must be a valid S3 bucket name"
+            )
+            raise ErrorWithStatus(msg, BlockedStatus)
+
+        # Ensure the bucket exists, creating it if missing and create_root_if_not_exists==True
+        s3_wrapper = S3BucketWrapper(
+            access_key=obj_storage_data["access-key"],
+            secret_access_key=obj_storage_data["secret-key"],
+            s3_service=f"{obj_storage_data['service']}.{obj_storage_data['namespace']}",
+            s3_port=obj_storage_data["port"],
+        )
+
+        if s3_wrapper.check_if_bucket_accessible(bucket_name):
+            return bucket_name
+        else:
+            if self.config["create_default_artifact_root_if_missing"]:
+                try:
+                    s3_wrapper.create_bucket(bucket_name)
+                    return bucket_name
+                except Exception as e:
+                    raise ErrorWithStatus(
+                        "Error with default S3 artifact store - bucket not accessible or "
+                        f"cannot be created.  Caught error: '{str(e)}",
+                        BlockedStatus,
+                    )
+            else:
+                raise ErrorWithStatus(
+                    "Error with default S3 artifact store - bucket not accessible or does not "
+                    "exist. Set create_default_artifact_root_if_missing=True to automatically "
+                    "create a missing default bucket",
+                    BlockedStatus,
+                )
```

Comment:
I'd split this method into two:

One for validating the name and access:

```python
def _validate_default_s3_bucket_name_and_access(self, bucket_name: str, s3_wrapper: S3BucketWrapper) -> bool:
  """Validates the default s3 bucket name is valid and the bucket is accessible.
        If it is not accessible and the `create_default_artifact_root_if_missing` config value
        is True, returns False; True otherwise.

         Args:
           bucket_name: ...
           s3_wrapper: ...
         Raises:
           ErrorWithStatus ...
             
  """
  ...
  if not validate_s3_bucket_name(self.config["default_artifact_root"]):
     raise ErrorWithStatus(...)

  if not s3_wrapper.check_if_bucket_accessible(...) and not self.config["create_default_artifact_root_if_missing"]:
    raise ErrorWithStatus("Bucket not accessible or does not exist...")
  elif not s3_wrapper.check_if_bucket_accessible(...) and self.config["create_default_artifact_root_if_missing"]:
    return False
  return True
```

Then have a dedicated method for creating the bucket:

```python
def _create_default_s3_bucket(self, s3_wrapper: S3BucketWrapper) -> None:
  """Creates an s3 bucket using the default_artifact_root config value.
        Raises:
          ErrorWithStatus: ...
  """
  try:
    s3_wrapper.create_bucket(self.config["default_artifact_root"])
  except:
    raise ErrorWithStatus(...)
```

Then you could use them in `_on_event()`:

```python
s3_wrapper = S3BucketWrapper(
            access_key=obj_storage_data["access-key"],
            secret_access_key=obj_storage_data["secret-key"],
            s3_service=f"{obj_storage_data['service']}.{obj_storage_data['namespace']}",
            s3_port=obj_storage_data["port"],
        )
if not self.validate_s3_bucket_name(bucket_name=self.config["default_artifact_root"], s3_wrapper=s3_wrapper):
   self._create_s3_default_buckets3_wrapper
default_artifact_root = self.config["default_artifact_rool"] # I believe we don't change this anywhere else, the name will always be this (?)
```

Reply:
Added your comment and fixed the tests :) thanks